### PR TITLE
Implement DSG changes to enable Auth Plugins

### PR DESCRIPTION
**Pull request by barshimi**
Issue Number: #3452

## PR Details
This PR is part of the plugins support process. I wrapped both createAuth function and createAdmin function with plugin wrapper, so it will support plugin access inside those functions. while doing so, I can manipulate/ change the final results of the DSG with plugins

## PR Checklist
- [ ] Tests for the changes have been added
- [ ] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

File: packages/amplication-code-gen-types/src/plugins-types.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+  entities: Entity[];
+  roles: Role[];
+  modules: Module[];
+  DTOs: DTOs;
+  plugins: PluginMap;
+  logger: winston.Logger;
+  utils: ContextUtil;
+}
+
+export type PluginWrapper = (args: any[], func: () => void) => any;
```
Comment (yuval-hazaz) [inline packages/amplication-code-gen-types/src/plugins-types.ts:26]: Is any[] still relevant? can we use types instead?

File: packages/amplication-code-gen-types/src/plugins-types.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+export interface DsgPlugin {
+  id?: string;
+  name?: string;
+  description?: string;
+  packageName: string;
+}
+
+export type PluginMap = {
+  [K in EventNames]?: {
+    before?: (<T>(context: DsgContext, params: any) => T)[];
```
Comment (yuval-hazaz) [inline packages/amplication-code-gen-types/src/plugins-types.ts:39]: can we use EventPrams instead of any?

File: packages/amplication-code-gen-types/src/plugins-types.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    after?: (<T>(context: DsgContext, params: any) => T)[];
```
Comment (yuval-hazaz) [inline packages/amplication-code-gen-types/src/plugins-types.ts:40]: we know that "after" is always required to return Modules[]

File: packages/amplication-code-gen-types/src/plugins-types.ts [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+    resource: string;
+    entityName: string;
+    entityType: string;
+    entityServiceModule: string;
+    entity: Entity;
+    srcDirectory: string;
+    extraMapping: { [key: string]: any };
+  };
+  after: Module[];
+}
```
Comment (yuval-hazaz) [inline packages/amplication-code-gen-types/src/plugins-types.ts:66]: because we wil have so many of these, it is better to use a separate file just for the events

File: packages/amplication-code-gen-types/src/plugins-types.ts [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+export interface CreateControllerModulesParams extends EventParams {
+  before: {
+    resource: string;
+    entityName: string;
+    entityType: string;
+    entityServiceModule: string;
+    entity: Entity;
+    srcDirectory: string;
+    extraMapping: { [key: string]: any };
```
Comment (yuval-hazaz) [inline packages/amplication-code-gen-types/src/plugins-types.ts:63]: no longer relevant

File: packages/amplication-data-service-generator/src/dsg-context.ts [hunk 6]
```diff
@@ -1,11 +1,26 @@
 import * as types from "@amplication/code-gen-types";
+import { ContextUtil } from "@amplication/code-gen-types";
+import { join } from "path";
+import { readStaticModules } from "./read-static-modules";
+import winston from "winston";
 
-export class DsgContext implements types.DsgContext {
+const contextUtil = {
+  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+  setDsgPath: (path: string) => join(__dirname, `../node_modules/${path}`),
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/dsg-context.ts:9]: We need to access "node_modules" because we are using `__dirname` in `dsg_context` via a wrapper function (setDsgPath).
If we will instruct the user to use __dirname directly in the plugin they will not need to access node_modules because the "static" folder is relative to their code

File: packages/amplication-code-gen-types/src/plugins-types.ts (hunk 3 above)
Comment (yuval-hazaz) [reply packages/amplication-code-gen-types/src/plugins-types.ts:40]: we also know that after will always send  Module[] as a second param

File: packages/amplication-code-gen-types/src/plugins-types.ts [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+  modules: Module[];
+  DTOs: DTOs;
+  plugins: PluginMap;
+  logger: winston.Logger;
+  utils: ContextUtil;
+}
+
+export type PluginWrapper = (args: any[], func: () => void) => any;
+
+export interface EventParams {}
```
[the remark this answers is not in the document]
Comment (yuval-hazaz) [reply packages/amplication-code-gen-types/src/plugins-types.ts:28]: we can even add a "fixed" after because it will always look the same 
(context: DsgContext, newModules: Module[]) : Module[];

File: packages/amplication-data-service-generator/src/plugin-wrapper.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    context.utils.skipDefaultBehavior = false;
+    if (!context.plugins.hasOwnProperty(event)) return func(args);
+
+    const beforePlugins = context.plugins[event]?.before;
+    const afterPlugins = context.plugins[event]?.after;
+
+    const beforeFuncResults = beforePlugins
+      ? pipe(...beforePlugins)(context, args)
+      : args;
+    const funcResults = await defaultBehavior(context, func, beforeFuncResults);
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/plugin-wrapper.ts:52]:
```suggestion
    const defaultBehaviourModules = await defaultBehavior(context, func, beforeFuncResults);
```

File: packages/amplication-data-service-generator/src/plugin-wrapper.ts [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+): Promise<any> => {
+  try {
+    const context = DsgContext.getInstance;
+    context.utils.skipDefaultBehavior = false;
+    if (!context.plugins.hasOwnProperty(event)) return func(args);
+
+    const beforePlugins = context.plugins[event]?.before;
+    const afterPlugins = context.plugins[event]?.after;
+
+    const beforeFuncResults = beforePlugins
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/plugin-wrapper.ts:49]:
```suggestion
    const updatedEventParams = beforePlugins
```

File: packages/amplication-data-service-generator/src/plugin-wrapper.ts [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+      ? pipe(...beforePlugins)(context, args)
+      : args;
+    const funcResults = await defaultBehavior(context, func, beforeFuncResults);
+
+    const afterFuncResults = afterPlugins
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/plugin-wrapper.ts:54]:
```suggestion
    const finalModules = afterPlugins
```

File: packages/amplication-data-service-generator/src/read-static-modules.ts [hunk 11]
```diff
@@ -18,7 +18,7 @@ export async function readStaticModules(
   const staticModules = await fg(`${directory}**/*`, {
     absolute: false,
     dot: true,
-    ignore: ["**.js", "**.js.map", "**.d.ts", "**/node_modules/**"],
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/read-static-modules.ts:21]: why did you remove this?

File: packages/amplication-data-service-generator/src/server/resource/controller/create-controller.ts [hunk 12]
```diff
@@ -112,6 +113,7 @@ export async function createControllerModules(
     WHERE_UNIQUE_INPUT: entityDTOs.whereUniqueInput.id,
 
     SWAGGER_API_AUTH_FUNCTION: getSwaggerAuthDecorationIdForClass(authProvider),
+    ...(extraMapping || {}),
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/resource/controller/create-controller.ts:116]: please remove

File: packages/amplication-data-service-generator/src/server/resource/controller/create-controller.ts [hunk 13]
```diff
@@ -54,18 +55,18 @@ const controllerBaseTemplatePath = require.resolve(
...
-  appInfo: AppInfo,
   resource: string,
   entityName: string,
   entityType: string,
   entityServiceModule: string,
   entity: Entity,
-  dtos: DTOs,
-  srcDirectory: string
+  srcDirectory: string,
+  extraMapping?: { [key: string]: any }
```
Comment (yuval-hazaz) [inline packages/amplication-data-service-generator/src/server/resource/controller/create-controller.ts:64]: please remove

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
outstanding work @barshimi 
please see my comments

File: packages/amplication-code-gen-types/src/plugins-types.ts (hunk 1 above)
Comment (yuval-hazaz) [reply packages/amplication-code-gen-types/src/plugins-types.ts:26]: use eventParams

File: packages/amplication-data-service-generator/src/dsg-context.ts (hunk 6 above)
Comment (yuval-hazaz) [reply packages/amplication-data-service-generator/src/dsg-context.ts:9]: remove the function

File: packages/amplication-data-service-generator/src/read-static-modules.ts (hunk 11 above)
Comment (yuval-hazaz) [reply packages/amplication-data-service-generator/src/read-static-modules.ts:21]: remove sicne we decided to remove the dsgPath

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

**Review by tupe12334 — APPROVED** [review]
